internal/integration/entity_synthesis.go

```diff
@@ -0,0 +1,169 @@
+// Package integration ...
+// Copyright 2021 New Relic Corporation. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+package integration
+
+import (
+	"strings"
+
+	sdk_metadata "github.com/newrelic/infra-integrations-sdk/v4/data/metadata"
+	"github.com/newrelic/nri-prometheus/internal/pkg/labels"
+)
+
+// Synthesizer group of rules to synthesis entities
+type Synthesizer struct {
+	rulesByConditions map[Condition]EntityRule
+}
+
+// NewSynthesizer initialize and return a Synthesizer from a set of EntityRules
+func NewSynthesizer(entitySynthesisDefinitions []SynthesisDefinition) Synthesizer {
+	s := Synthesizer{
+		rulesByConditions: make(map[Condition]EntityRule),
+	}
+	for _, ed := range entitySynthesisDefinitions {
+		ed.tagRules = ed.Tags.getTagRules()
+		s.addConditions(ed.EntityRule)
+
+		for _, er := range ed.Rules {
+			er.EntityType = ed.EntityType
+			er.tagRules = append(er.Tags.getTagRules(), ed.EntityRule.tagRules...)
+			s.addConditions(er)
+		}
+	}
+	return s
+}
+func (s *Synthesizer) addConditions(rule EntityRule) {
+	if !rule.isValid() {
+		return
+	}
+	for _, c := range rule.Conditions {
+		s.rulesByConditions[c] = rule
+	}
+}
+
+// SynthesisDefinition contains rules to synthesis entities from metrics
+type SynthesisDefinition struct {
+	EntityRule `mapstructure:",squash"`
+	Rules      []EntityRule `mapstructure:"rules"`
+}
+
+// EntityRule contains rules to synthesis an entity
+type EntityRule struct {
+	EntityType string      `mapstructure:"type"`
+	Identifier string      `mapstructure:"identifier"`
+	Name       string      `mapstructure:"name"`
+	Conditions []Condition `mapstructure:"conditions"`
+	Tags       Tags        `mapstructure:"tags"`
+	tagRules   []tagRule
+}
+
+func (rule EntityRule) isValid() bool {
+	return rule.Identifier != "" && rule.Name != "" && rule.EntityType != ""
+}
+
+// Condition contains parameters used to match entities from metrics
+type Condition struct {
+	Attribute string `mapstructure:"attribute"`
+	Prefix    string `mapstructure:"prefix"`
+	Value     string `mapstructure:"value"`
+}
+
+// match evaluates the condition for a particular existing attribute value.
+func (c Condition) match(attribute string) bool {
+	if c.Value != "" {
+		return c.Value == attribute
+	}
+	// this returns true if c.Prefix is "" and is ok since the attribute exists
+	return strings.HasPrefix(attribute, c.Prefix)
+}
+
+// Tags key value attributes
+type Tags map[string]map[string]interface{}
+
+type tagRule struct {
+	name          string
+	entityTagName string
+}
+
+func (t Tags) getTagRules() []tagRule {
+	var tr []tagRule
+	for k, v := range t {
+		if v == nil {
+			tr = append(tr, tagRule{name: k, entityTagName: k})
+			continue
+		}
+		if newName, ok := v["entityTagName"]; ok {
+			newNameS, _ := newName.(string)
+			tr = append(tr, tagRule{name: k, entityTagName: newNameS})
+		}
+	}
```

Comment:
Similarly as above, I think short names here hurt readability, specially `k` and `v` (since it's not super obvious what the keys and values of this map actually model).